docs(fern): rewrite key terminology glossary (#1500) (#2355)
NVIDIA-NeMo/Gym: fern/versions/latest/pages/about/concepts/key-terminology.mdx

---
title: "Key Terminology"
description: "Essential vocabulary for evaluation, model training, RL workflows, and NeMo Gym."
---

Essential vocabulary for evaluation, model training, RL workflows, and NeMo Gym. You'll encounter these terms throughout the tutorials and documentation.

Deeper treatment: [Environments](/about/concepts/environments), [Evaluation](/about/concepts/evaluation), [Training](/about/concepts/training), [Architecture](/about/architecture).

## Overview

A **dataset** is a JSONL file of **tasks** (one row = one problem). Running the agent on a task produces a **rollout** (also called a **trajectory**) — one attempt and its record.

Each run follows:

1. **seed_session** — set up isolated state for this attempt  

> 2. **agent loop** — call the model, use tools, repeat until done  

Reviewer: ```suggestion
2. **agent loop** — call the LLM-model (policy-model), use tools, repeat until done  
```

Author: Addressed with modality-neutral wording: “call the model (policy)” rather than LLM-specific.

3. **verify** — score the attempt → **reward**

Two compositions matter:

| | Built from | Notes |
| --- | --- | --- |
| **Agent** | **Model** + **Agent harness** | The harness turns model calls into a loop with tools and stop conditions. |

> | **Environment** | **Dataset** + **Agent harness** + **Verifier** + **State** | Everything needed to run and score a task. The **model is outside** the environment. |

Reviewer: One important distinction here is that the **Agent harness** is not a responsibility of the Environment
it is part of the Agent as you described in the line above. But the Environment packaging that we are doing contains a reference to the Agent Server

| **Environment** | **Dataset** + **Agent harness** + **Verifier** + **State** | Everything needed to run and score a task. The **model is outside** the environment. |

Author: Addressed: Environment is now `Dataset + Verifier + State` (+ env tools). Explicit note that the **harness belongs to the Agent**; Gym configs may still *reference* an agent server for packaging only.

> The **model** is the LLM you call — local weights or a remote API endpoint. In Gym it is usually exposed by the **Model server** (or your harness can call an endpoint directly). The same environment can be used as a **benchmark** (fixed eval) or a **training environment** (learn from rewards / synthetic data).

Reviewer 2: nit: may be a vlm, cf super

Author: Went with modality-neutral wording instead of listing LLM/VLM: **model / policy** is what you call for generation (local weights or remote endpoint). That covers text, vision-language, and other modalities without baking a modality into the glossary.

> The **model** is the LLM you call — local weights or a remote API endpoint. In Gym it is usually exposed by the **Model server** (or your harness can call an endpoint directly). The same environment can be used as a **benchmark** (fixed eval) or a **training environment** (learn from rewards / synthetic data).

Reviewer: ```suggestion
The **model** is the LLM you call — local weights or a remote API endpoint. In Gym it is usually exposed by the **Model server** (or your harness can call an endpoint directly). The same environment can be used rollout generation generation on a **benchmark** (fixed eval) or used for **policy-model training** (learn from rewards / synthetic data).
```

Author: Addressed in spirit (benchmark vs training usage), with modality-neutral **model / policy** wording instead of LLM-only.

| Concept | NeMo Gym component |
| --- | --- |

> | Dataset | JSONL (`responses_create_params` + `verifier_metadata`) — [Prepare Data](/data) |

Reviewer 2: I think calling this a component is confusing. I would put the link to the data section, and the function names

Author: Addressed: retitled the map to “In NeMo Gym” and described Dataset as JSONL row fields + [Prepare Data](/data) (not as a server “component”).

| Agent harness | **Agent server** (`responses_api_agents/`) — [Agent Server](/agent-server) |

> | Verifier + state (+ env tools) | **Resources server** (`resources_servers/`) — [Build Verifiers](/build-verifiers) |

Reviewer 2: this is confusing to me. the state is obtained by a snapshot of the environment, no?

Author: Clarified: per-attempt **state** is the mutable session world inside the resources server (`seed_session` → tools mutate it → `verify()`), not a snapshot of the whole environment config. Mapping table updated accordingly.

| Model | **Model server** (`responses_api_models/`) — wraps local or remote LLM endpoints — [Model Server](/model-server) |
| Protocol | **Responses API** (Chat Completions converted via middleware when needed) |

## Glossary

### Tasks & rollouts

| Term | Definition |
| --- | --- |
| **Task** | One problem for the agent to solve — one dataset row (`responses_create_params` + `verifier_metadata`). |
| **Dataset** | Collection of tasks plus metadata needed for scoring (usually JSONL). See [Prepare Data](/data). |

> | **Rollout / trajectory** | **Rollout** (verb): execute a policy in an environment — take actions and record what happens. **Rollout** (noun) / **trajectory**: the ordered record of one attempt (states, actions, rewards). In Gym these names are aliases; architecture calls the resulting trajectory a *rollout*. Multiple rollouts per task support metrics like pass@k. See [trajectory capabilities](/reference/trajectory-capabilities). |

Reviewer: ```suggestion
| **Rollout / trajectory** | **Rollout** (verb): execute an agent in an environment — take actions and record what happens. **Rollout** (noun) / **trajectory**: the ordered record of one attempt (states, actions, rewards). In Gym these names are aliases; architecture calls the resulting trajectory a *rollout*. Multiple rollouts per task support metrics like pass@k. See [trajectory capabilities](/reference/trajectory-capabilities). |
```

Author: Applied: rollout (verb) is now “execute an agent in an environment”.

| **Task attempt** | One rollout for a specific task. Multiple attempts per task capture different approaches and support pass@k. |
| **Trace** | Debug-oriented log of a rollout (timing, tool I/O, metadata) beyond the scored trajectory. |
| **Rollout batch** | Multiple rollouts generated together (across tasks for throughput, or grouped on one task for methods like GRPO). |
| **Rollout collection** | Running inference, tools, and verification at scale to produce scored rollouts. Start with the [Quickstart](/get-started/quickstart) or [Evaluate](/evaluation). |

### Environment

| Term | Definition |
| --- | --- |

> | **Environment** | Everything required for an agent to complete a task except the model: `Dataset + Agent harness + Verifier + State`. See [Environments](/about/concepts/environments) and [Build Environments](/environment-tutorials). |

Reviewer: Same recommendation as before.

Author: Addressed: Environment glossary row matches Overview — harness is a packaging reference, not part of the Environment formula.

| **Agent** | Model + agent harness. |
| **Multi-turn** | Dialogue across turns where conversation context (and often state) persists. |
| **Multi-step** | Sequential tool calls or intermediate steps in the agent loop before completion (may be single- or multi-turn). See [Multi-Step Environment](/environment-tutorials/multi-step-environment). |
| **Tool use** | Function calling — invoking external capabilities (APIs, code execution, databases, and so on). |

### Evaluation

| Term | Definition |
| --- | --- |
| **Evaluation** | Run an agent on tasks, score results, and measure performance. See [Evaluation](/about/concepts/evaluation) and the [Evaluate](/evaluation) workflow. |
| **Benchmark** | Repeatable evaluation built on an environment: fixed dataset, metrics, and comparison protocol. Not every environment is a benchmark. See [Benchmarks](/evaluation/benchmarks). |
| **pass@k** | Fraction of tasks with at least one success among *k* rollouts. See [Aggregate Metrics](/evaluation/aggregate-metrics). |

### Training

| Term | Definition |
| --- | --- |

> | **Training environment** | Environment used to improve a model — as an online RL reward signal, or to generate synthetic training data. See [Training](/about/concepts/training). |

Reviewer: not sure if there is need to make a distinction between a "Training environment" from an Environment apart from the taskset and data contamination

Author: Agreed — dropped “Training environment” as a separate type. Overview now says the same environment shape is used for benchmark eval vs training; differences are mainly task split / contamination controls.

| **Demonstration data** | SFT examples from successful (high-reward) rollouts. |
| **Preference pairs** | DPO data: same task, high-reward vs low-reward rollouts (preferred vs dispreferred). |
| **SFT** | Supervised fine-tuning from examples of good behavior (demonstration data). |
| **RL** | Improve the policy through environment interaction and reward signals. |
| **Online / offline** | **Online**: update the policy from rewards while interacting (e.g. GRPO). **Offline**: train from pre-collected rollouts (e.g. SFT, DPO). See [Offline training with rollouts](/tutorials/training-tutorials/offline-training-w-rollouts). |

> | **DPO** | Offline preference training from pairs of rollouts. |

Reviewer 2: nit: i would expand the acronyms

Reviewer 3: +1

Author: Addressed in `5cd0537` / earlier review commit: expanded **SFT (Supervised Fine-Tuning)**, **RL (Reinforcement Learning)**, **DPO (Direct Preference Optimization)**, and **GRPO (Group Relative Policy Optimization)**.

Author: Addressed — full names are expanded in the Training glossary.
